OAuth: Support mapping generic OAuth users to Grafana organizations

treydock opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/master/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the master branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

-->

**What this PR does / why we need it**:
Add `orgs_attribute_path` to generic oauth configuration similar to `role_attribute_path` but for mapping a user to one or many Grafana organizations.

**Which issue(s) this PR fixes**:

<!--

* Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #23473  and would potentially make #22605 a non-issue for me.

**Special notes for your reviewer**:

Currently using `auto_assign_org_id` doesn't work with OAuth when the desire is to have people login in to an org that's set with `auto_assign_org_id`, and then assigning them after the fact to a different org (#22605). To have maximum flexibility of putting people in multiple organizations I opted to make the values comma delimited.

treydock committed:
OAuth: Support mapping generic oauth users to Grafana organizations

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=23661) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=23661) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=23661) it.</sub>

marefr commented:
@treydock thanks. We need to discuss whether this is something we want to accept now/short term. Will provide an update here when we're done with that.

marefr commented:
@treydock we have discussed this and came to the conclusion that this is something we could accept and would make sense since we support that for LDAP. However, for us to accept ut we would like to see support for all OAuth providers (generic oauth, github, azure ad, octa, google, gitlab).

One possible drawback with your proposed solution is that it limits same org role to all mapped organizations. Do you have any input on this?

We have Grafana 7.0 release coming up and won't have time to review this until after. But let us know if you have any questions/concerns or want input on something.

treydock commented:
> One possible drawback with your proposed solution is that it limits same org role to all mapped organizations. Do you have any input on this?

I think one way around this and keeping minimal amount of changes would maybe be to have the value for mapping contain more data so something like this:

```
orgs_attribute_path = contains(info.groups[*], 'staff') && 1:Editor,2:Viewer
```

So the above would map staff to Editor in orgId=1 and Viewer in orgId=2. The approach taken by LDAP is the most flexible but is a bigger change I would think as might have to incorporate a toml file or some other file to have many items listed that can't be defined with basic INI settings.
 
> We have Grafana 7.0 release coming up and won't have time to review this until after. But let us know if you have any questions/concerns or want input on something.

That's fine, if this can't make it into 7.x after 7.0 then hopefully can make it into 8.0.

marefr commented:
> I think one way around this and keeping minimal amount of changes would maybe be to have the value for mapping contain more data so something like this:
> 
> ```
> orgs_attribute_path = contains(info.groups[*], 'staff') && 1:Editor,2:Viewer
> ```
> 
> So the above would map staff to Editor in orgId=1 and Viewer in orgId=2. The approach taken by LDAP is the most flexible but is a bigger change I would think as might have to incorporate a toml file or some other file to have many items listed that can't be defined with basic INI settings.

With this approach either `role_attribute_path` or `orgs_attribute_path` will be used which is not optimal. Can be quite cumbersome to configure as well if you have more than a couple organizations. Maybe reuse `role_attribute_path`, but maybe a bit weird. Alternatively deprecate `role_attrbiute_path` and introduce `role_mapping_path` like 

`role_mapping_path = contains(info.groups[*], 'staff') && [{ "role": "Editor", "orgId": [1] }, { "role": "Viewer", "orgId": [ 2 ] } ]`

This can be very cumbersome to configure correctly. 

Maybe something like this:
```ini
[auth.generic_oauth.group_mapping]
attribute_path = contains(info.groups[*], 'staff') && true
role = Editor
org_id = 1

[auth.generic_oauth.group_mapping]
attribute_path = contains(info.groups[*], 'staff2') && true
role = Viewer
org_id = 2
```

Not super proud of this either and not sure it will work. Maybe toml is the way to go?

> That's fine, if this can't make it into 7.x after 7.0 then hopefully can make it into 8.0.

Yes

treydock commented:
I think the multiple sections could work.  I think I see how it could be implemented.  Should I update this pull request with the multiple sections approach? Given the code I see for OAuth that seems like the easiest to implement without overly complicating a single config option.

marefr commented:
@treydock you can try it out/update PR if you want so that we get a feel for how such a solution would look and feel.

treydock committed:
Initial support for generic oauth group mappings

treydock committed:
Support setting Grafana Admin with generic OAuth

treydock commented:
I've added my initial pass at supporting the `auth.generic_oauth.group_mapping` sections. If this approach looks good I can add support to the other oauth providers and further update documentation.  I'll also remove `role_attribute_path` in favor of this new section style if no major changes needed to this approach.

treydock committed:
Lint fix

treydock committed:
Avoid overwriting orgID assignment of so first one is one used, similar to LDAP

treydock commented:
In my testing I discovered that multiple sections in INI with same name doesn't work but it is supported by the INI go library but requires newer version: https://github.com/go-ini/ini/commit/faf04e23c2215448b87cd9fd510d3e06445a219f.  I am testing that change now.

treydock committed:
Upgrade ini dependency and support loading duplicate INI sections

treydock commented:
I've updated with my attempt at doing multiple sections with same name and it's not working and I've not yet figured out why. I added some debug logging and get this when Grafana starts up:

```
t=2020-04-28T16:27:41+0000 lvl=dbug msg="GroupMappings parsed section=auth.github.group_mapping count=1"
t=2020-04-28T16:27:41+0000 lvl=dbug msg="GroupMappings parsed section=auth.gitlab.group_mapping count=1"
t=2020-04-28T16:27:41+0000 lvl=dbug msg="GroupMappings parsed section=auth.google.group_mapping count=1"
t=2020-04-28T16:27:41+0000 lvl=dbug msg="GroupMappings parsed section=auth.generic_oauth.group_mapping count=1"
t=2020-04-28T16:27:41+0000 lvl=dbug msg="GroupMappings parsed section=auth.grafananet.group_mapping count=1"
t=2020-04-28T16:27:41+0000 lvl=dbug msg="GroupMappings parsed section=auth.grafana_com.group_mapping count=1"
t=2020-04-28T16:27:41+0000 lvl=dbug msg="GroupMappings parsed section=auth.azuread.group_mapping count=1"
t=2020-04-28T16:27:41+0000 lvl=dbug msg="GroupMappings parsed section=auth.okta.group_mapping count=1"
```

I only defined sections for `auth.generic_oauth.group_mappings`:

```
[auth.generic_oauth.group_mapping]
role_attribute_path = contains(groups[*], 'sysstf') && 'Admin'
org_id = 1
grafana_admin = true

[auth.generic_oauth.group_mapping]
role_attribute_path = contains(groups[*], 'oscall') && 'Editor' || 'Viewer'
org_id = 1

[auth.generic_oauth.group_mapping]
role_attribute_path = contains(groups[*], 'sysstf') && 'Admin'
org_id = 2

[auth.generic_oauth.group_mapping]
role_attribute_path = 'Viewer'
org_id = 2
```

Commit: https://github.com/treydock/grafana/commit/9a4bccc900311478542010bde5ea2cfe0c89b2a5

treydock commented:
I have read the change for ini library and it looks like multiple sections should work but not having much luck getting it to work. I can try toml next unless someone sees something wrong with how I tried to support INI and non-unique sections.

treydock committed:
Revert "Upgrade ini dependency and support loading duplicate INI sections"

This reverts commit 9a4bccc900311478542010bde5ea2cfe0c89b2a5.

treydock committed:
Use TOML to define group mapping configs

treydock committed:
Fix assignment of grafana admin

treydock commented:
@marefr I switched this pull request to use TOML for the mappings.  I wasn't sure if it made since to instead of using the references of `group_mappings` in configs to maybe call it `claims_mappings` or `role_org_mappings`.

Did some manual verification using Docker instance of Grafana pointed at Keycloak for OAuth.

```
[auth.generic_oauth]
...
group_mappings_file = /etc/grafana/oauth.toml
```

oauth.toml:

```
[[group_mappings]]
role_attribute_path = "contains(groups[*], 'sysstf') && 'Admin'"
org_id = 1
grafana_admin = true

[[group_mappings]]
role_attribute_path = "contains(groups[*], 'oscall') && 'Editor' || 'Viewer'"
org_id = 1

[[group_mappings]]
role_attribute_path = "'Viewer'"
org_id = 2
```

treydock commented:
If this approach is acceptable I can expand it to include all the OAuth providers and not just generic_oauth and also remove unnecessary `role_attribute_path`.

treydock committed:
Remove commented out code

treydock commented:
This is ready for review.

treydock commented:
Checking in to avoid stale close.

bergquist commented:
related: https://github.com/grafana/grafana/issues/24588

mapshen commented:
@marefr since 7.0 is now out, would it be possible for us to have this? Thanks!

treydock commented:
Any chance of getting a review of this PR?

chhetripradeep commented:
Hello grafana maintainers, is it possible to review this PR please, pretty sure this will unblock many folks.

marefr commented:
@treydock sorry, been crazy busy lately. Will prioritize giving this a review this week.

marefr commented:
I have started review/testing of this, but some things came in between. Sorry for that. Expect a review this week

nohupped commented:
@marefr Is it possible to kindly include the PR treydock/grafana/pull/1 along with this while you review? This is to extend the Oauth user mapping to Okta provider, and will unblock our Grafana-Okta integration.

Thank you very much.

marefr commented:
@nohupped Looks like this PR makes your PR obsolete? Cannot include it in this, please open a new PR in this repository if it does something else than what this provide..

nohupped commented:
Thank you @marefr . This is to extend this mapping to the Okta provider, and the changes are in `pkg/login/social/okta_oauth.go`. I will open a new PR once this is reviewed and merged.

marefr commented:
@nohupped this PR is not complete. If we accept it the idea is to add this support to all OAuth providers and generic OAuth.

nohupped commented:
> @nohupped this PR is not complete. If we accept it the idea is to add this support to all OAuth providers and generic OAuth.

Thank you @marefr for your reply. 

I have only updated it for the Okta Oauth provider because our project was depending on this. If you find this approach acceptable, I can try to implement this for the rest of the Oauth providers and submit it as a new PR once the current PR is reviewed and merged.

marefr commented:
@nohupped Okta is included in `all OAuth providers and generic OAuth` so wait and see until this PR is merged.

nohupped commented:
Sure @marefr, While I tested with this PR earlier, enabling the `[auth.generic_oauth]` section worked as expected and retained my access to other org IDs, but while I enabled the `[auth.okta]` section, it didn't work and I lost access to other Org ids. Apologies if I missed something here. 

I will wait for this to be merged, will test again and will re-open a new one if it still persists. Thank you very much for your input.

marefr commented on pkg/login/social/social.go:
```diff
@@ -96,12 +102,20 @@ func NewOAuthService() {
 			continue
 		}
 
+		groupMappingsConfig, err := readConfig(info.GroupMappingsFile)
```
Should only read config if group mapping file defined (`info.GroupMappingsFile != ""`). Right now checking out this code I get a panic on startup since returned `groupMappingsConfig` is nil and `info.GroupMappings = groupMappingsConfig.GroupMappings` panics - why I think it's easier to check if a file is provided here and only then read the config, handle error or assign group mappings.

But lets discuss this when my other comments/suggestions are fixed.

treydock replied:
I went ahead and made this change.

marefr commented on pkg/login/social/social.go:
```diff
@@ -96,12 +102,20 @@ func NewOAuthService() {
 			TlsClientKey:       sec.Key("tls_client_key").String(),
 			TlsClientCa:        sec.Key("tls_client_ca").String(),
 			TlsSkipVerify:      sec.Key("tls_skip_verify_insecure").MustBool(),
+			GroupMappingsFile:  sec.Key("group_mappings_file").String(),
```
Don't need to add it to this struct since you only use it below. Init `GroupMappings` per default instead.

```suggestion
			GroupMappings: []setting.OAuthGroupMapping{},
```

treydock replied:
This change was committed

marefr commented on pkg/login/social/social.go:
```diff
@@ -96,12 +102,20 @@ func NewOAuthService() {
 
+		groupMappingsConfig, err := readConfig(info.GroupMappingsFile)
+		if err != nil {
+			log.Error(3, "Error reading group mappings config: %s", err.Error())
```
```suggestion
			oauthLogger.Error("Failed to read group mappings file", "provider", name, "error", err)
```

treydock replied:
This change was committed, only different is I kept `file` as one of the items logged just so logs show which file failed to be read.

marefr commented on pkg/login/social/social.go:
```diff
@@ -174,6 +188,7 @@ func NewOAuthService() {
 				emailAttributeName:   info.EmailAttributeName,
 				emailAttributePath:   info.EmailAttributePath,
 				roleAttributePath:    info.RoleAttributePath,
+				groupMappings:        info.GroupMappings,
```
If/when adding this support to all OAuth providers we shouldn't need to do this. Instead groupMappings could possibly be added to `SocialBase` since `newSocialBase` takes `info` as argument. SocialBase could then hold the `extractGroupMappings` method and be used by all OAuth providers.

treydock replied:
I can add this change when I extend this PR to support the other OAuth providers, if that's something this PR should do.

marefr commented on pkg/login/social/social.go:
```diff
@@ -223,3 +238,35 @@ var GetOAuthProviders = func(cfg *setting.Cfg) map[string]bool {
+	}
+	result := &setting.OAuthGroupMappingsConfig{}
+
+	oauthLogger.Info("OAuth enabled, reading config file", "file", configFile)
```
```suggestion
	oauthLogger.Debug("Reading group mappings file", "file", configFile)
```

treydock replied:
This change is done.

marefr commented on pkg/login/social/social.go:
```diff
@@ -223,3 +238,35 @@ var GetOAuthProviders = func(cfg *setting.Cfg) map[string]bool {
 }
+
+func readConfig(configFile string) (*setting.OAuthGroupMappingsConfig, error) {
+	if configFile == "" {
+		return nil, nil
+	}
```
As commented earlier, suggest to move the logic if group mappings file specified before calling `readConfig`

treydock replied:
This change is made

marefr commented on pkg/login/social/social.go:
```diff
@@ -223,3 +238,35 @@ var GetOAuthProviders = func(cfg *setting.Cfg) map[string]bool {
+
+	oauthLogger.Info("OAuth enabled, reading config file", "file", configFile)
+
+	fileBytes, err := ioutil.ReadFile(configFile)
+	if err != nil {
+		return nil, errutil.Wrap("Failed to load OAuth config file", err)
+	}
+
+	_, err = toml.Decode(string(fileBytes), result)
```
Think you can replace this with `_, err = toml.DecodeFile(configFile, result)`

treydock replied:
Change made

marefr commented on pkg/login/social/social.go:
```diff
@@ -223,3 +238,35 @@ var GetOAuthProviders = func(cfg *setting.Cfg) map[string]bool {
+
+	_, err = toml.Decode(string(fileBytes), result)
+	if err != nil {
+		return nil, errutil.Wrap("Failed to load OAuth config file", err)
```
```suggestion
		return nil, errutil.Wrap("Failed to load OAuth group mappings file", err)
```

treydock replied:
Change made

marefr commented on pkg/login/social/social.go:
```diff
@@ -223,3 +238,35 @@ var GetOAuthProviders = func(cfg *setting.Cfg) map[string]bool {
+	// Validate role_attribute_path is set
+	for _, groupMapping := range result.GroupMappings {
+		if groupMapping.RoleAttributePath == "" {
+			return nil, xerrors.New("Failed to validate role_attribute_path setting")
```
```suggestion
			return nil, xerrors.New("OAuth group mapping require role_attribute_path to be set")
```

treydock replied:
This change is made

marefr commented on pkg/login/social/social_test.go:
```diff
@@ -0,0 +1,30 @@
+
+func TestReadingSimpleGroupMappingSettings(t *testing.T) {
+	config, err := readConfig("testdata/simple.toml")
+	assert.Nil(t, err, "No error when reading oauth config")
```
```suggestion
	require.NoError(t, err)
```

marefr commented on pkg/login/social/social_test.go:
```diff
@@ -0,0 +1,30 @@
+func TestReadingSimpleGroupMappingSettings(t *testing.T) {
+	config, err := readConfig("testdata/simple.toml")
+	assert.Nil(t, err, "No error when reading oauth config")
+	count := len(config.GroupMappings)
+	assert.EqualValues(t, 1, count)
+	if count != 1 {
+		return
+	}
+	assert.EqualValues(t, "role", config.GroupMappings[0].RoleAttributePath)
```
```suggestion
	require.Len(t, config.GroupMappings, 1)
	require.Equal(t, "role", config.GroupMappings[0].RoleAttributePath)
```

treydock replied:
This change is done.

marefr commented on pkg/login/social/social_test.go:
```diff
@@ -0,0 +1,30 @@
+
+func TestReadingFullGroupMappingSettings(t *testing.T) {
+	config, err := readConfig("testdata/full.toml")
+	assert.Nil(t, err, "No error when reading oauth config")
+	count := len(config.GroupMappings)
+	assert.EqualValues(t, 3, count)
+	if count != 3 {
+		return
+	}
+	assert.EqualValues(t, "contains(info.groups[*], 'admin') && 'Admin'", config.GroupMappings[0].RoleAttributePath)
+	assert.EqualValues(t, true, *config.GroupMappings[0].IsGrafanaAdmin)
```
```suggestion
	require.NoError(t, err)
	require.Len(t, config.GroupMappings, 3)
	require.Equal(t, "contains(info.groups[*], 'admin') && 'Admin'", config.GroupMappings[0].RoleAttributePath)
	require.True(t, config.GroupMappings[0].IsGrafanaAdmin)
```

treydock replied:
Only difference I had to make here was keeping `IsGrafanaAdmin` as pointer, per other comment.

marefr commented on pkg/login/social/social_test.go:
```diff
@@ -0,0 +1,30 @@
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
```
Use `require` instead to make the test fail directly instead of continue as when using `assert`.
```suggestion
	"github.com/stretchr/testify/require"
```

treydock replied:
Updated

marefr commented on pkg/login/social/generic_oauth.go:
```diff
@@ -239,6 +249,29 @@ func (s *SocialGenericOAuth) extractRole(data *UserInfoJson) (string, error) {
 
+func (s *SocialGenericOAuth) extractGroupMappings(data *UserInfoJson) ([]setting.OAuthGroupMapping, error) {
+	var groupMappings []setting.OAuthGroupMapping
+	if s.groupMappings == nil {
```
Given my comment regarding default init in social.go we could remove this if statement.

treydock replied:
Removed this condition

marefr commented on pkg/setting/setting_oauth.go:
```diff
@@ -17,6 +17,19 @@ type OAuthInfo struct {
+	RoleAttributePath string `toml:"role_attribute_path"`
+	Role              string
+	OrgId             int   `toml:"org_id"`
+	IsGrafanaAdmin    *bool `toml:"grafana_admin"`
```
Should need pointer since false is the default if not provided.

```suggestion
	IsGrafanaAdmin    bool `toml:"grafana_admin"`
```

treydock replied:
I ran into a lot of problems when testing this when not using a pointer.  I could only get the Grafana Admin assignment to work when I used a pointer.  I stumbled across that solution when looking at how the LDAP code assigned Grafana Admin.  I think it might have to do with what `extUser.IsGrafanaAdmin` expects `*bool` for `IsGrafanaAdmin`:

```go
type ExternalUserInfo struct {
	OAuthToken     *oauth2.Token
	AuthModule     string
	AuthId         string
	UserId         int64
	Email          string
	Login          string
	Name           string
	Groups         []string
	OrgRoles       map[int64]RoleType
	IsGrafanaAdmin *bool // This is a pointer to know if we should sync this or not (nil = ignore sync)
	IsDisabled     bool
}
```

treydock replied:
Would is possibly work to make `IsGrafanaAdmin` from TOML not a pointer and then assign by doing this: `extUser.IsGrafanaAdmin = &groupMapping.IsGrafanaAdmin`, or should I just keep things a pointer?

marefr replied:
Sorry, you're right. My bad, let's keep the pointer

marefr reviewed (changes requested):
Great work. This is looking promising and using external TOML file make sense to me. Have added some minor comments/suggestions, please double check them in case I put in a typo somewhere :).

Lets try and fix these things. Then I can take this for a spin. And then we can discuss ideas for adding this to all OAuth providers.

Again, sorry it took quite a while.

treydock committed:
Improvements from review

treydock committed:
More improvements from review

marefr commented:
@treydock great :+1: Build is currently failing, could you please merge with latest grafana master branch? Please don't rebase since it'll messes up the history of reviews etc.

treydock committed:
Merge remote-tracking branch 'upstream' into oauth-orgs

dskatz commented:
@treydock  I think this is fantastic.  I have been using a patch for while to integrate this type of functionality so I am super excited to see this has gained traction.  I just have one question related to this PR.

Coming from the LDAP toml, it was possible to have a configuration where a user could be in both group 'Everyone' and group "Team", where Everyone would have the 'Viewer', and Team, "Editor" or "Admin".  

I submitted a [PR](https://github.com/grafana/grafana/pull/26240/files#diff-e44496d49035224a35512ba7a59f7caeR208) to outline the logic I am referring to.

treydock commented:
@marefr I've merged latest master.

@dskatz I'm not sure I follow what you're trying to achieve.  If someone is Admin they are inertially already Editor but I don't think I follow what you're trying to achieve. In case it helps here's an example config I was using during test:

```
[[group_mappings]]
role_attribute_path = "contains(groups[*], 'sysstf') && 'Admin'"
org_id = 1
grafana_admin = true

[[group_mappings]]
role_attribute_path = "contains(groups[*], 'oscall') && 'Editor' || 'Viewer'"
org_id = 1

[[group_mappings]]
role_attribute_path = "'Viewer'"
org_id = 2
```

Group `sysstf` got assigned Admin to orgId=1, `oscall` gets assigned Editor and everyone else is Viewer. Then for orgId=2 everyone gets Viewer.  In the case of my claims I am a member of both `sysstf` and `oscall` but the first config block was used so I was assigned Admin with Grafana Admin privileges despite being in both groups, the first matching TOML section is the one used for a given org mapping.  The line you linked maybe also done here: https://github.com/treydock/grafana/blob/oauth-orgs/pkg/api/login_oauth.go#L210-L215

dskatz commented:
@treydock, thanks for the clarification.  It is based on the order in the toml that prevents a permission for a user from being downgraded.  That should work perfectly for our use-cases as well.

marefr commented:
@treydock please merge with master again

@dskatz not sure what's the idea is with your PR. In case you think this PR doesn't cover a certain use case that you need please explain it here so that we can possibly include it in this PR. I don't see the point of opening an additional PR?

treydock committed:
Merge remote-tracking branch 'upstream/master' into oauth-orgs

treydock commented:
@marefr I've merged in master

marefr commented:
@treydock this looks good :+1: 

Speaking about enable this for all OAuth providers there are a couple of things we need to discuss to come up with a single solution. 

Github and Azure AD for example extracts groups via HTTP request and/or claims compared to generic oauth which extracts role from token/user info payload. To support org/role mapping in this scenario we could convert the already extracted groups to JSON and apply your current mapping solution here. Bad thing with this is that you have to write JMESPath expression in the mappings on a JSON structure that's "made up" But could be solved with documentation I guess. TOML for GitHub and Azure AD could then look something like this:

```toml
[[group_mappings]]
role_attribute_path = "contains(groups, 'admin') && 'Admin'"
org_id = 1
grafana_admin = true

[[group_mappings]]
role_attribute_path = "contains(groups, 'editor') && 'Editor'"
org_id = 1

[[group_mappings]]
role_attribute_path = "contains(groups, 'admin') && 'Admin'"
org_id = 2
```

An alternative to this would be if generic oauth allowed groups to be extracted and then the mapping would refer to extracted groups instead. Something like this (TOML):
```toml
# extract all groups
groups_attribute_path = "info.groups[*]"

[[group_mappings]]
groups = [ "admin" ]
role = "Admin"
org_id = 1
grafana_admin = true

[[group_mappings]]
groups = [ "editor" ]
role = "Editor"
org_id = 1

[[group_mappings]]
groups = [ "admin" ] 
role = "Admin"
org_id = 2

[[group_mappings]]
groups = [ "*" ] 
role = "Viewer"
org_id = 2
```

We're going to discuss this internally as well. For now, do you have any thoughts/feedback around this?

treydock commented:
@marefr The first set of examples looks very similar to what I was already testing with, how does it differ exactly than what's already been coded, is it the lack of `[*]` in the JMESPath expression? I think either solution would work really though the second set of examples might be easier to work with for people configuring Grafana and looks a little cleaner.

aknuds1 commented on docs/sources/auth/generic-oauth.md:
```diff
@@ -172,11 +172,19 @@ allowed_organizations =
 To ease configuration of a proper JMESPath expression, you can test/evaluate expressions with custom payloads at http://jmespath.org/.
 
-### Role mapping
+### Role and Organization mapping
```
```suggestion
### Role and organization mapping
```

aknuds1 replied:
Headings should be [sentence case](https://github.com/grafana/grafana/blob/master/contribute/style-guides/documentation-style-guide.md#headings).

aknuds1 commented on docs/sources/auth/generic-oauth.md:
```diff
@@ -172,11 +172,19 @@ allowed_organizations =
-### Role mapping
+### Role and Organization mapping
+
+Defining role and organization mapping is handled by setting a path to TOML mapping file via `group_mappings_file`, example:
```
```suggestion
Defining role and organization mapping is handled by setting a path to a TOML mapping file via `group_mappings_file`, for example:
```

aknuds1 commented on docs/sources/auth/generic-oauth.md:
````diff
@@ -172,11 +172,19 @@ allowed_organizations =
+
+Defining role and organization mapping is handled by setting a path to TOML mapping file via `group_mappings_file`, example:
+
+```bash
````
```suggestion
```ini
```

aknuds1 commented on docs/sources/auth/generic-oauth.md:
```diff
@@ -172,11 +172,19 @@ allowed_organizations =
 **Basic example:**
 
-In the following example user will get `Editor` as role when authenticating. The value of the property `role` will be the resulting role if the role is a proper Grafana role, i.e. `Viewer`, `Editor` or `Admin`.
+In the following example user will get `Editor` as role in OrgID `2` when authenticating. The value of the property `role` will be the resulting role if the role is a proper Grafana role, i.e. `Viewer`, `Editor` or `Admin`.
```
```suggestion
In the following example, the user will get `Editor` as role in OrgID `2` when authenticating. The value of the property `role` will be the resulting role if the role is a proper Grafana role, i.e. `Viewer`, `Editor` or `Admin`.
```

aknuds1 commented on docs/sources/auth/generic-oauth.md:
````diff
@@ -189,12 +197,14 @@ Payload:
 
 Config:
 ```bash
````
```suggestion
```toml
```

aknuds1 commented on docs/sources/auth/generic-oauth.md:
```diff
@@ -189,12 +197,14 @@ Payload:
 **Advanced example:**
 
-In the following example user will get `Admin` as role when authenticating since it has a group `admin`. If a user has a group `editor` it will get `Editor` as role, otherwise `Viewer`.
+In the following example user will get `Admin` as role in Orgs `1` and `2` when authenticating since it has a group `admin`. If a user has a group `editor` it will get `Editor` as role for Org `1`, otherwise `Viewer` in Org `2`.
```
The original wording doesn't really make sense to me, I try to make sense of it through the following wording:

```suggestion
In the following example, the user will get `Admin` as role in organizations `1` and `2` when authenticating if they have a group `admin`. If, on the other hand, the user has a group `editor`, they will get `Editor` as role for organization `1. If the user isn't in the `admin` group, they will get the `Viewer` role in organization `2`.
```

aknuds1 commented on docs/sources/auth/generic-oauth.md:
````diff
@@ -214,5 +224,14 @@ Payload:
 
 Config:
 ```bash
````
```suggestion
```toml
```

aknuds1 commented on pkg/login/social/generic_oauth.go:
```diff
@@ -239,6 +249,26 @@ func (s *SocialGenericOAuth) extractRole(data *UserInfoJson) (string, error) {
+			continue
+		}
+		if role == "" {
+			s.log.Debug(fmt.Sprintf("role_attribute_path did not produce a role: %s", mapping.RoleAttributePath))
```
Quoting the value is preferable since it may contain whitespace:
```suggestion
			s.log.Debug(fmt.Sprintf("role_attribute_path did not produce a role: %q", mapping.RoleAttributePath))
```

aknuds1 commented on pkg/login/social/social.go:
```diff
@@ -223,3 +240,27 @@ var GetOAuthProviders = func(cfg *setting.Cfg) map[string]bool {
+
+	_, err := toml.DecodeFile(configFile, result)
+	if err != nil {
+		return nil, errutil.Wrap("Failed to load OAuth group mappings file", err)
```
Error strings should not be [capitalized](https://github.com/golang/go/wiki/CodeReviewComments#error-strings):
```suggestion
		return nil, errutil.Wrap("failed to load OAuth group mappings file", err)
```

aknuds1 reviewed (commented):
Some initial comments from my side, until I've analyzed the proposed scheme further.

aknuds1 commented on pkg/login/social/generic_oauth_test.go:
```diff
@@ -332,3 +333,177 @@ func TestUserInfoSearchesForEmailAndRole(t *testing.T) {
+				ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+					w.WriteHeader(http.StatusOK)
+					w.Header().Set("Content-Type", "application/json")
+					_, _ = io.WriteString(w, string(response))
```
```suggestion
					_, err := w.Write(response)
					require.NoError(t, err)
```

aknuds1 commented on pkg/login/social/generic_oauth_test.go:
```diff
@@ -332,3 +333,177 @@ func TestUserInfoSearchesForEmailAndRole(t *testing.T) {
+		for _, test := range tests {
+			provider.groupMappings = test.GroupMappings
+			t.Run(test.Name, func(t *testing.T) {
+				response, _ := json.Marshal(test.APIURLReponse)
```
```suggestion
				response, err := json.Marshal(test.APIURLReponse)
				require.NoError(t, err)
```

aknuds1 commented on pkg/login/social/social.go:
```diff
@@ -223,3 +240,27 @@ var GetOAuthProviders = func(cfg *setting.Cfg) map[string]bool {
+	}
+
+	if len(result.GroupMappings) == 0 {
+		return nil, xerrors.New("OAuth enabled but no group mappings defined in config file")
```
```suggestion
		return nil, fmt.Errorf("OAuth enabled but no group mappings defined in config file")
```

aknuds1 commented on pkg/login/social/social.go:
```diff
@@ -223,3 +240,27 @@ var GetOAuthProviders = func(cfg *setting.Cfg) map[string]bool {
+	// Validate role_attribute_path is set
+	for _, groupMapping := range result.GroupMappings {
+		if groupMapping.RoleAttributePath == "" {
+			return nil, xerrors.New("OAuth group mapping require role_attribute_path to be set")
```
```suggestion
			return nil, fmt.Errorf("OAuth group mapping require role_attribute_path to be set")
```

aknuds1 commented on pkg/login/social/social.go:
```diff
@@ -6,21 +6,25 @@ import (
 
+	"github.com/BurntSushi/toml"
 	"golang.org/x/oauth2"
+	"golang.org/x/xerrors"
```
```suggestion
```

treydock committed:
Update docs/sources/auth/generic-oauth.md

Co-authored-by: Arve Knudsen <arve.knudsen@gmail.com>

treydock committed:
Update docs/sources/auth/generic-oauth.md

Co-authored-by: Arve Knudsen <arve.knudsen@gmail.com>

treydock committed:
Update docs/sources/auth/generic-oauth.md

Co-authored-by: Arve Knudsen <arve.knudsen@gmail.com>

treydock committed:
Update docs/sources/auth/generic-oauth.md

Co-authored-by: Arve Knudsen <arve.knudsen@gmail.com>

treydock committed:
Update docs/sources/auth/generic-oauth.md

Co-authored-by: Arve Knudsen <arve.knudsen@gmail.com>

treydock committed:
Update docs/sources/auth/generic-oauth.md

Co-authored-by: Arve Knudsen <arve.knudsen@gmail.com>

treydock committed:
Update docs/sources/auth/generic-oauth.md

Co-authored-by: Arve Knudsen <arve.knudsen@gmail.com>

treydock committed:
Update pkg/login/social/generic_oauth.go

Co-authored-by: Arve Knudsen <arve.knudsen@gmail.com>

treydock committed:
Update pkg/login/social/social.go

Co-authored-by: Arve Knudsen <arve.knudsen@gmail.com>

treydock committed:
Update pkg/login/social/generic_oauth_test.go

Co-authored-by: Arve Knudsen <arve.knudsen@gmail.com>

treydock committed:
Update pkg/login/social/generic_oauth_test.go

Co-authored-by: Arve Knudsen <arve.knudsen@gmail.com>

treydock committed:
Update pkg/login/social/social.go

Co-authored-by: Arve Knudsen <arve.knudsen@gmail.com>

treydock committed:
Update pkg/login/social/social.go

Co-authored-by: Arve Knudsen <arve.knudsen@gmail.com>

treydock committed:
Update pkg/login/social/social.go

Co-authored-by: Arve Knudsen <arve.knudsen@gmail.com>

treydock committed:
Fix missing import

aknuds1 commented:
@torkelo says we should hold off on eventually merging this, until we've given it/orgs enough thought, just to make that clear.

treydock commented:
@aknuds1 Anything I can help on with regard to thought on orgs or is that more of an internal discussion?

aknuds1 commented:
It's an internal discussion @treydock, we basically need to give it some time and thought.

torkelo commented:
I am very sorry that we have given you mixed signals on this feature contribution. Things have happened since. We entertained the idea to deprecate multi orgs support (https://github.com/grafana/grafana/issues/24588) but after the huge negative feedback we decided to keep multi orgs support.

However to in order to make that decision we plan to limit new multi orgs features. This feature falls into that space. Really sorry that this impacted work already started that we encouraged.

But I do think https://github.com/grafana/grafana/issues/22605 is a real issue that we need to fix.

RyanW8 commented:
It would be nice to see support for this feature @torkelo. we're starting to embrace Grafana orgs for multi-tenancy and the inability to set orgs dynamically from oAuth is a real hinderance.

l13t commented:
We're also interested in using organizations in Grafana because we need to limit access to datasources in our Grafana instance.

xenofree commented:
We're also interested for this feature

oddlittlebird commented:
Please do not comment on closed PRs. There are more appropriate routes for feedback. @achatterjee-grafana , I recommend closing discussion on this PR.
